### ISSUE TITLE
Usable Helm Chart

### ISSUE THREAD

**Issue by stevehipwell** [backstage/backstage#4945]
## Feature Suggestion

The Backstage Helm chart should be able to deploy a usable version of Backstage with dynamic plugin support and pre-built Docker images.

This chart should be published to the [Artifact Hub](https://artifacthub.io/).

The current Helm chart isn't fit for this purpose, although I don't think it was even intended to be.

## Possible Implementation

- Use an init container to download the plugins
- Template the plugin config from _values.yaml_ into the correct location(s)

## Context

I'm happy to work on this if it's deemed technically possible.

**Comment by Crevil** [comment]
We had some discussion in #4429 that covers some of the same problems. The plugins are not plug and play items that can be dropped into the code base but requires all sorts of integrations in the front and backend applications. I have a hard time seeing how we could make this happen. 😞

**Comment by stevehipwell** [comment]
Thanks for pointing me at #4429 @Crevil, I'd missed that. I am slightly confused as to why it's been closed as an issue when an operator seems like the logical way to support deploying Backstage if plugins require customisation at the point of integration?

Given the complexity around plugins it looks like we'd need a Backstage K8s operator responsible for managing deployments; this operator could then be deployed via a Helm chart. At a MVP level the operator could define a CRD for a Backstage instance and a CRD for a plugin installation; the plugin CRD define an API to support the per plugin customization required.

**Comment by Crevil** [comment]
The main reason for closing that issue was no concrete use case was identified.

See https://github.com/backstage/backstage/issues/4429#issuecomment-784423600 for details around a thought experiment around modelling this.

If you can identity a concrete use case where the operator pattern makes sense, please fill in a comment on the issue (we can always open it up again).

**Comment by stevehipwell** [comment]
Was a RFC around this ever created?

**Comment by freben** [comment]
I don't know of one offhand.

**Comment by stevehipwell** [comment]
I think this is still a valid issue to have open.

**Comment by benjdlambert** [comment]
@stevehipwell so I think to be honest we should be closing this ticket in favour of the current helm chart that exists today.

> dynamic plugin support 

Is a long way off, and Backstage is right now, architected in a way that this isn't possible without configuration of the JSX and the code to customise it to your needs. Maybe one day we will have that in a dynamic way, but either way I don't think that this ticket is the right place to track that. 

Thoughts?

**Comment by jacek-jablonski** [comment]
Correctly designed charts would certainly be very useful in deploying backstage. At first, it would be enough to extend and streamline the current Charts in such a way that you can expand and make the configuration more flexible.

At this point, for example, without using quite ugly hacks, it is not possible to securely put a password to the Postgres database (a reference to an existing Kubernetes Secret created in a more secure way, e.g. using ExternalSecret).

To sum up, your own docker images would still be required to start, but the Charts should allow for more flexibility (please look at Bitnami's Charts as an example of good flexibility).

**Comment by Rugvip** [comment]
What are some advantages of us providing charts over using the built-in k8s resources directly? Is it a deployment environment constraint or something to make it easier to set up? I'm having a hard time seeing the charts actually being easier to use than the plain k8s objects in https://backstage.io/docs/deployment/k8s.

**Comment by jacek-jablonski** [comment]
It's definitely easier to deploy the initial version with just a single helm command. Just to see the results and take an initial trip around the tool. Later you can simply adjust and tune some configurations to make production deployment without reinventing the wheel. Upgrading the whole deployment is also easier when some elements are added to the system or redesigned.

When you just want to look around, test some stuff, check integrations if it's what you are looking for, it is way more harder and time-consuming to do that with the plain k8s objects (potentially a chance for more teams adopting backstage).

Moreover, it is already a bit of a standard that almost every piece of software that can be installed on Kubernetes has a helm chart which is at least a very correct way of deployment.

**Comment by Rugvip** [comment]
Would it make sense to switch around how configuration is done in the existing chart so that the user of the chart configures an entire app-config YAML instead of each individual value? That seems to me like the biggest thorn in the current charts. Then there could be a few extra switches like whether to include Postgres or not, but overall keep it quite slim

**Comment by jacek-jablonski** [comment]
@Rugvip yes, exactly.

Moreover there are two cases not covered by current charts:
- Requiring Postgres CA Secret to existing DB instance (https://github.com/backstage/backstage/blob/master/contrib/chart/backstage/templates/postgresql-ca-config.yaml#L18-L20). Shouldn't `.Values.backend.postgresCertMountEnabled` be taken into account? When using SQL Proxy encryption to postgres is provided with proxy, so it should not be hardly enforced.
- Requiring password to existing postgres to be placed in `values.yaml` https://github.com/backstage/backstage/blob/master/contrib/chart/backstage/templates/postgresql-password-backend-secret.yaml#L14. There should be an option to define custom secret with postgres password as we wouldn't like to place  passwords in raw form in `values.yaml` (due to GitOps).

I would be grateful for taking into consideration the above comments. Thanks.

**Comment by freben** [comment]
@martina-if Would you like to chip in on this discussion?

**Comment by dominik-bln** [comment]
Since we just deployed a test installation with the current Helm chart, I'd really like to see this reopened. The experience was pretty painful to say the least, as we ran into the following problems:

- Ingress resources have hard-coded `nginx.ingress.kubernetes.io/configuration-snippet` which is disabled in our cluster because of https://github.com/kubernetes/ingress-nginx/issues/7837 and honestly should be up to the user whether STS is enabled or not
- It is not possible to use the Chart without a TLS secret. It is not documented that this is needed and since everything else is not production ready, seems superfluous. Common practice is to have a switch to enable this.
- The default configuration provided in the docs does not work because `frontend.enabled` is `false` by default, so you are greeted with 503 service unavailable. In the Ingress logs we then saw `Error obtaining Endpoints for Service "devops/dev-portal-backstage-frontend": no object matching key "devops/dev-portal-backstage-frontend" in local store`
- We didn't set the Lighthouse configuration at first since we thought we don't really need this in a first trial. Through the defaults this creates an ingress resource with `demo.example.com` as a TLS host, giving `Unexpected error validating SSL certificate "devops/dev-portal-backstage-tls" for server "demo.example.com": x509: certificate is valid for: ...`, so you really have to set the same hostname six times.
- Seeing the default docker images coming from the personal account of a user as `martinaif/backstage-k8s-demo-backend` is a bit discouraging for a CNCF project.
- There is no Helm repo, this adds friction in the beginning.

Having this Helm Chart as our first encounter with Backstage, is to be honest not a great start. In the current state, this is barely usable.

I'd be happy to contribute some improvements, but currently the direction seems unclear to me since this issue is closed and some [other](https://github.com/backstage/backstage/issues/9248#issuecomment-1026587828) [comments](https://github.com/backstage/backstage/issues/9248#issuecomment-1061119895) suggest that a Helm install is not considered that important. For a CNCF project this approach is a bit unexpected. It should definitely be possible to provide a great out-of-the-box experience with a Helm chart that still allows customization by building your own Docker images.

**Comment by dominik-bln** [comment]
This was just a test install, but the plan for us would be to get something production ready deployed if the test goes well. For kubernetes deploys we mainly do them from CI pipelines and then often have a staging environment to test updates. In my experience Helm charts are usually the most convenient way to get multiple production ready instances with minimal configuration differences maintained in the long run so we started with the chart. Often the Helm charts easily take care of persistent volumes, secret management, maybe security and network policies and simplify updates. We also usually use the Helm `--wait` feature so that the pipeline fails in case an update goes wrong (there are ways to also get this with done with `kubectl`, but less convenient).

With manifest files we'd have to do the templating ourselves or commit the templates for staging and production and ensure that they are kept in sync. Additionally we'd need to compare them regularly with the docs or upstream changes to see whether something important has changed (with a versioned Helm chart, we can usually get updates notifications via renovate bot).

I see that having a versioned Helm chart that is not just for testing is an additional burden on the maintainers, but for me as a user the convenience of having that would be significant and a one-line Helm install command is also a great "selling point" to other users in my view.

**Comment by jurschel** [comment]
I have to agree with @dominik-bln.  I am amazed that these helm chart issues are still not worked out.  For instance right now the chart doesn't install the front-end anymore if you follow the instructions on the helm page...  I'm not talking about a production instance either.. I mean, just a chart that will let you get something up inside of 10m to look at it.  One of the great things about having a helm chart is the speed of deployment isn't it?  I looked a year ago and thought well the community will work it out and I'll come back and this will just work...  It is apparently so hard to get a production instance of backstage running a company has sprouted up telling people to just not even try https://www.opslevel.com/.

**Comment by ChrisJBurns** [comment]
I can attest, the first thing I do when installing anything on a Cluster now is look for the official Helm Chart for it. I haven't tried the Backstage Helm Chart yet because I noticed that it needs `nginx` to be installed, which I won't work for us as we use Istio Gateway because of the seamless integration with its Service Mesh.

Is the recommendation to instead use a bunch of Kubernetes Manifests to install Backstage? We use FluxCD so that isn't the end of the world, but I would still prefer a Helm Chart as the whole point is the abstraction it gives you away from the Kubernetes manifests.

**Comment by Crevil** [comment]
I actually think this is one of the main reasons for not providing a helm chart for this. As everyone has their own way of doing ingress and all other sorts of things, it is hard to abstract that part away. 

Taking that out what you git is a`Deployment` manifest, which, to me at least, is simple enough to just create.

As Backstage doesn't require anything Kubernetes, it could just as well be deployed to any other service which makes for the chart to maintained by those in need of it. So far, it seems, has no one though steped up for this task, that is to maintain such a chart.

**Comment by ChrisJBurns** [comment]
I will have to look at how to deploy it via Kubernetes manifests first. Once I become familiar with it, I'll try and see if I can put it into a Helm Chart. I've recently done the `pact-foundation/pact-broker` Helm Chart that I'm awaiting merging because I noticed they didn't have one. So I don't mind doing it. However, I need to look at how much of it can be configurable first and also what the pre-reqs are - if any.

**Comment by Crevil** [comment]
The chart is currently located in a `contrib/chart` directory but there are no mention of it in the README I can see 👍 

Maybe @freben or @Rugvip has opinions on this?

**Comment by vinzscam** [comment]
Hi all 👋,

We heard your feedback and we are in the process of rethinking/rewriting the current chart.

We have a prototype here https://github.com/vinzscam/backstage-chart looking for your feedback, to understand whether this is the right direction for building a good Backstage Helm chart.

A couple of highlights:
- There is no pre-defined docker image ready to be used, so you have to come with your own image
- The customization of the plugins need to be done within your Backstage application
- The chart creates a service and a Deployments: additional resources like postgres database are not included in the chart, so that you can choose what’s the best for your use case (deploy additional resources within the cluster, or use external managed solutions)
- It is possible to pass environment variables, secrets together with external `app-config.yaml` (defined through `ConfigMap`)

Looking forward to hearing your feedback!

**Comment by stevehipwell** [comment]
@vinzscam once this is released would you like a PR to support running a _DEV_ database in Kubernetes so the chart can be used for quick POCs?

**Comment by ChrisJBurns** [comment]
Great news! Am currently on holiday but can contribute to this when I'm back. I've done some stuff around feature flagging Postgres databases so even if we did offer it as part of the chart, it would be entirely config driven.

**Comment by vinzscam** [comment]
> @vinzscam once this is released would you like a PR to support running a _DEV_ database in Kubernetes so the chart can be used for quick POCs?

Isn't already possible to do this using the postgres chart and including the database url in the app-config.xml?

**Comment by stevehipwell** [comment]
@vinzscam I was thinking of adding the postgres chart as a sub chart and make it optional. I do this for my charts which need postgres so I can run chart testing on them and people can kick the tyres quickly.

**Comment by dominik-bln** [comment]
@vinzscam To me the general direction you describe sounds very reasonable and keeps options open. The only thing for my personal usage I'm missing is an `ingress` manifest. I think it would be possible to add this in a generic enough way and toggle it with `ingress.enabled` as seems to be a pattern in other charts. This is mainly a matter of convenience, I can write an ingress manifest on my own, but without it my deploy must run two commands or I need to somehow adjust the chart before the deploy.

Regarding the Postgres subchart install I think the question to ask is: What use-cases should the chart serve?

1. Quick test install
2. Production ready install
3. both

If it's only 2., then the subchart is additional complexity, but for 1. or 3. it would make sense because it could make it easier for people to get started if they are not so experienced.

**Comment by vinzscam** [comment]
> @vinzscam Do you mind if I add some issues / PR's to the [Helm Chart repo](https://github.com/vinzscam/backstage-chart) you linked above? This saves me creating my own repository as I'm just about to write up a Helm chart for our Backstage deployment. So if I can use yours, we could possible make it something that people will find useful.

sure, it would be amazing!

**Comment by martina-if** [comment]
Hi, I am not sure what the opinion of the maintainers is but I'd be more than happy to replace the existing charts with something better like what @vinzscam is doing.

I'm happy to deprecate the existing charts and remove them in a couple of weeks. Opinions?

**Comment by ChrisJBurns** [comment]
Hi @martina-if if the chart that @vinzscam has becomes something people are happy with as a base, then I'm happy to deprecate the old ones. It actually works for me (with a few tweaks - although I shall contribute these tweaks) as opposed to the existing ones.

**Comment by cmoulliard** [comment]
The new chart looks promising but documentation should be improved to explain how to build backstage to expose the frontend and backend as currently only the backend is running - https://github.com/halkyonio/backstage/blob/2ce2bd3c5e1d1a5057fc4dbe2405ee7f0af9dd8b/README.md#instructions
and you will got an error 503 if you try to access the ingress route

**Comment by kevinbutto** [comment]
hi! any updates on the progress for this? thanks :)

**Comment by ChrisJBurns** [comment]
@kevinbutto  Yep, we have a Helm Chart for Backstage that currently works (I am using it). https://github.com/vinzscam/backstage-chart. It's a basic version of it, but should work.

**Comment by froblesmartin** [comment]
> @kevinbutto Yep, we have a Helm Chart for Backstage that currently works (I am using it). https://github.com/vinzscam/backstage-chart. It's a basic version of it, but should work.

@ChrisJBurns is the idea to make that one official once it is a little bit more developed?

**Comment by ChrisJBurns** [comment]
@froblesmartin  I believe so. @vinzscam  Can confirm :+1:

**Comment by cmoulliard** [comment]
I think that it is time that the community work on this ticket as an alternative project has been developed by [red hat ](https://github.com/redhat-developer/helm-backstage) and that would be great to unify the different efforts

**Comment by ChrisJBurns** [comment]
@cmoulliard  I think gathering around the repository that @vinzscam  has is probably best - as this is the one that may be taken forward and merged into the main backstage org.

**Comment by sabre1041** [comment]
id be happy to work with @vinzscam and the community on a more canonical Helm chart that could be used by the broader community!

**Comment by riginoommen** [comment]
https://developers.redhat.com/articles/2022/10/24/red-hat-joins-backstageio-community

**Comment by marrws** [comment]
Hello @vinzscam, sorry to ping you directly

Given the recent deprecation notice in vinzscam/backstage-chart@0d864f1

Is backstage/charts now considered the official helm chart?

**Comment by Rugvip** [comment]
Yep, [backstage/charts](https://github.com/backstage/charts) is now the place to got for charts.

I'll close this issue as we'll be moving over to only use those charts, please open any new issues in that repo instead.
